File: v4.8/source/includes/indexes/indexes.py
# start-index-single
movies.create_index("title")
# end-index-single

# start-index-single-query
query = { "title": "Batman" }
sort = [("title", 1)]

cursor = movies.find(query).sort(sort)
# end-index-single-query

# start-compound-index
movies.create_index([("type", pymongo.ASCENDING), ("genre", pymongo.ASCENDING)])
# end-compound-index

# start-index-compound-query
query = { "type": "movie", "genre": "Drama" }
sort = [("type", pymongo.ASCENDING), ("genre", pymongo.ASCENDING)]

cursor = movies.find(query).sort(sort)
# end-index-compound-query

# start-index-multikey
result = movies.create_index("cast")
# end-index-multikey

# start-index-multikey-query
query = { "cast": "Viola Davis" }

cursor = movies.find(query)
# end-index-multikey-query

# start-index-text-single
movies.create_index(
    [( "plot", "text" )]
)
# end-index-text-single

# start-index-text-single-query
query = { "$text": { "$search": "a time-traveling DeLorean" } }

cursor = movies.find(query)
# end-index-text-single-query

# start-index-text-multi
result = myColl.create_index(
    [("title", "text"), ("genre", "text")],
    default_language="english",
    weights={ "title": 10, "genre": 3 }
)
# end-index-text-multi

# start-index-geo
theaters.create_index(
    [( "location.geo", "2dsphere" )]
)
# end-index-geo

# start-index-wildcard
movies.create_index({ "location.$**": pymongo.ASCENDING })
# end-index-wildcard

# start-index-unique
theaters.create_index("theaterId", unique=True)
# end-index-unique

# start-index-clustered
sample_mflix.create_collection("movies", clusteredIndex={
    "key": { "_id": 1 }, 
    "unique": True
})
# end-index-clustered

# start-remove-index
movies.drop_index("_title_")
# end-remove-index

# start-create-search-index
index = {
    "definition": {
        "mappings": {
            "dynamic": True
        }
    },
    "name": "<index name>",
}

collection.create_search_index(index)
# end-create-search-index

# start-create-vector-search-index

from pymongo.operations import SearchIndexModel

search_index_model = SearchIndexModel(
  definition={
    "fields": [
      {
        "type": "vector",
        "numDimensions": <number of dimensions>,
        "path": "<field to index>",
        "similarity":  "<select from euclidean, cosine, dotProduct>"
      }
    ]
  },
  name="<index name>",
  type="vectorSearch",
)

collection.create_search_index(model=search_index_model)

# end-create-vector-search-index

# start-create-search-indexes

search_idx = SearchIndexModel(
    definition ={
        "mappings": {
            "dynamic": True
        }
    },
    name="my_index",
)

vector_idx = SearchIndexModel(
  definition={
    "fields": [
      {
        "type": "vector",
        "numDimensions": <number of dimensions>,
        "path": "<field to index>",
        "similarity": "<select from euclidean, cosine, dotProduct>"
      }
    ]
  },
  name="my_vector_index",
  type="vectorSearch",
)

indexes = [search_idx, vector_idx]

collection.create_search_indexes(models=indexes)
# end-create-search-indexes

# start-list-search-indexes
results = list(collection.list_search_indexes())

for index in results:
    print(index)
# end-list-search-indexes

# start-update-search-indexes
new_index_definition = {
    "mappings": {
        "dynamic": False
    }
}

collection.update_search_index("my_index", new_index)
# end-update-search-indexes

# start-update-vector-search-indexes
new_index_definition = {
    "fields": [
        {
            "type": "vector",
            "numDimensions": 1536,
            "path": "<field to index>",
            "similarity": "euclidean"
        },
    ]
}

collection.update_search_index("my_vector_index", new_index_definition)
# end-update-vector-search-indexes

# start-delete-search-indexes
collection.drop_search_index("my_index")
# end-delete-search-indexes

File: content/kotlin-sync/v5.0/source/includes/indexes/indexes.kt
import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.model.Indexes
import com.mongodb.client.model.Filters
import com.mongodb.client.model.Sorts
import com.mongodb.client.model.SearchIndexModel
import com.mongodb.kotlin.client.MongoClient
import org.bson.codecs.pojo.annotations.BsonId
import org.bson.types.ObjectId
import org.bson.Document

// start-movie-class
data class Movie(
    @BsonId
    val id: ObjectId,
    val title: String? = "",
    val type: String? = "",
    val genres: List<String>? = null,
    val cast: List<String>? = null,
    val plot: String? = "",
)
// end-movie-class

fun main() {
    val uri = "<connection string URI>"

    val settings = MongoClientSettings.builder()
        .applyConnectionString(ConnectionString(uri))
        .retryWrites(true)
        .build()

    val mongoClient = MongoClient.create(settings)
    val database = mongoClient.getDatabase("sample_mflix")
    val collection = database.getCollection<Movie>("movies")

    // start-remove-index
    collection.dropIndex("_title_")
    // end-remove-index

    // start-remove-all-indexes
    collection.dropIndexes()
    // end-remove-all-indexes

    // start-index-single
    collection.createIndex(Indexes.ascending(Movie::title.name))
    // end-index-single

    // start-index-single-query
    val filter = Filters.eq(Movie::title.name, "Batman")
    val sort = Sorts.ascending(Movie::title.name)
    val results = collection.find(filter).sort(sort)

    results.forEach { result ->
        println(result)
    }
    // end-index-single-query

    // start-index-compound
    collection.createIndex(Indexes.ascending(Movie::type.name, Movie::genres.name))
    // end-index-compound

    // start-index-compound-query
    val filter = and(
        eq(Movie::type.name, "movie"),
        `in`(Movie::genres.name, "Drama")
    )
    val sort = Sorts.ascending(Movie::type.name, Movie::genres.name)
    val results = collection.find(filter).sort(sort)

    results.forEach { result ->
        println(result)
    }
    // end-index-compound-query
    
    // start-create-search-index
    val index = Document("mappings", Document("dynamic", true))
    collection.createSearchIndex("<index name>", index)
    // end-create-search-index

    // start-create-search-indexes
    val indexOne = SearchIndexModel("<first index name>", Document("mappings", Document("dynamic", true)))
    val indexTwo = SearchIndexModel("<second index name>", Document("mappings", Document("dynamic", true)))
    collection.createSearchIndexes(listOf(indexOne, indexTwo))
    // end-create-search-indexes

    // start-list-search-indexes
    val results = collection.listSearchIndexes()

    results.forEach { result ->
        println(result)
    }
    // end-list-search-indexes

    // start-update-search-indexes
    val newIndex = Document("mappings", Document("dynamic", true))
    collection.updateSearchIndex("<index to update>", newIndex)
    // end-update-search-indexes

    // start-drop-search-index
    collection.dropIndex("<index to delete>")
    // end-drop-search-index
}
